File: solutions/erlang/diamond/1/src/diamond.erl
-module(diamond).

-export([rows/1]).

rows(Letter) ->
  TopRows = make_top_rows(Letter),
  BottomRows = make_bottom_rows(Letter),
  lists:append([TopRows, BottomRows]).

make_top_rows(LastLetter)
  when (LastLetter == "A") ->
  ["A"];

make_top_rows(LastLetter) ->
  Letters = lists:seq(hd("A"), hd(LastLetter)),
  lists:map(make_row(LastLetter), Letters).

make_bottom_rows(LastLetter)
  when (LastLetter == "A") ->
  [];

make_bottom_rows(LastLetter) ->
  Letters = lists:seq(hd("A"), hd(LastLetter) - 1),
  lists:map(make_row(LastLetter), lists:reverse(Letters)).

make_row(LastLetter) ->
  fun(CurrentLetter) ->
    OuterSpaces = create_outer_spaces(hd(LastLetter) - CurrentLetter),
    InnerSpaces = create_inner_spaces(CurrentLetter - hd("A")),
    Row = create_row(OuterSpaces, CurrentLetter, InnerSpaces),
    unicode:characters_to_list(Row)
  end.

create_row(OuterSpaces, CurrentLetter, _InnerSpaces)
  when (CurrentLetter == hd("A")) ->
  [
    OuterSpaces,
    CurrentLetter,
    OuterSpaces
  ];

create_row(OuterSpaces, CurrentLetter, InnerSpaces) ->
  [
    OuterSpaces,
    CurrentLetter,
    InnerSpaces,
    CurrentLetter,
    OuterSpaces
  ].

create_outer_spaces(DistanceFromA) ->
  SpacesNeeded = lists:seq(1, DistanceFromA),
  create_spaces(SpacesNeeded).

create_inner_spaces(DistanceFromA)
  when (DistanceFromA == 0) ->
  "";

create_inner_spaces(DistanceFromA)
  when (DistanceFromA == 1) ->
  " ";

create_inner_spaces(DistanceFromA) ->
  SpacesNeeded = lists:seq(1, (DistanceFromA * 2) - 1),
  create_spaces(SpacesNeeded).

create_spaces(SpacesNeeded) ->
  Spaces = lists:map(fun(_X) -> " " end, SpacesNeeded),
  unicode:characters_to_binary(Spaces).

File: solutions/elixir/diamond/1/lib/diamond.ex
defmodule Diamond do
  @doc """
  Given a letter, it prints a diamond starting with 'A',
  with the supplied letter at the widest point.
  """
  @spec build_shape(char) :: String.t()
  def build_shape(letter) do
    diamond =
      List.flatten([
        make_top_rows(letter),
        make_bottom_rows(letter)
      ])
      |> Enum.join("\n")

    diamond <> "\n"
  end

  defp make_top_rows(last_letter)
       when last_letter === ?A,
       do: ["A"]

  defp make_top_rows(last_letter),
    do:
      ?A..last_letter
      |> Enum.map(&make_row(&1, last_letter))

  defp make_bottom_rows(last_letter)
       when last_letter === ?A,
       do: []

  defp make_bottom_rows(last_letter),
    do:
      ?A..(last_letter - 1)
      |> Enum.reverse()
      |> Enum.map(&make_row(&1, last_letter))

  defp make_row(current_letter, last_letter) do
    Enum.join(
      create_row(
        create_outer_spaces(last_letter - current_letter),
        current_letter,
        create_inner_spaces(current_letter - ?A)
      )
    )
  end

  defp create_outer_spaces(distance_from_last_letter)
       when distance_from_last_letter === 0,
       do: ""

  defp create_outer_spaces(distance_from_last_letter),
    do: create_spaces(1..distance_from_last_letter)

  defp create_inner_spaces(distance_from_a)
       when distance_from_a === 0,
       do: ""

  defp create_inner_spaces(distance_from_a)
       when distance_from_a === 1,
       do: " "

  defp create_inner_spaces(distance_from_a),
    do: create_spaces(1..(distance_from_a * 2 - 1))

  defp create_spaces(spaces_needed),
    do: Enum.map(spaces_needed, &to_space/1)

  defp to_space(_count),
    do: " "

  defp create_row(outer_spaces, current_letter, _inner_spaces)
       when current_letter === ?A,
       do: [outer_spaces, stringify(current_letter), outer_spaces]

  defp create_row(outer_spaces, current_letter, inner_spaces),
    do: [
      outer_spaces,
      stringify(current_letter),
      inner_spaces,
      stringify(current_letter),
      outer_spaces
    ]

  defp stringify(ascii_code),
    do: List.to_string([ascii_code])
end
